---
rule: Z001
title: Function names should be camelCase
enabled: true
---

# Z001: Function names should be camelCase

Zig convention is to use `camelCase` for function names. This distinguishes functions from types (which use `PascalCase`) and variables (which use `snake_case`).

Functions that return types are an exception - see [Z005](Z005.md) for type function naming.

## Examples

### Bad

```zig
// expect: Z001
fn MyFunction() void {}
```

```zig
// expect: Z001
fn my_function() void {}
```

### Good

```zig
fn myFunction() void {}
```

```zig
fn myHTTPHandler() void {}
```

Leading underscores are allowed for unused parameters or internal functions:

```zig
fn _internalHelper() void {}
```

## Rationale

Consistent naming helps readers quickly identify what an identifier represents:
- `camelCase` → function or method
- `PascalCase` → type or type-returning function
- `snake_case` → variable or field

This matches the style used in the Zig standard library.
