---
rule: Z006
title: Variable names should be snake_case
enabled: true
---

# Z006: Variable names should be snake_case

Variable and constant names should use snake_case, unless they are type aliases.

## Examples

### Bad

```zig
// expect: Z006
const myValue = 42;
```

```zig
// expect: Z006
var currentIndex = 0;
```

### Good

```zig
const my_value = 42;
```

```zig
var current_index = 0;
```

Type aliases are allowed to be PascalCase:

```zig
const MyType = struct { value: u32 };
```

```zig
const Allocator = std.mem.Allocator;
```

Function aliases are allowed to be camelCase:

```zig
fn fooBar() void {}
const myAlias = fooBar;
```

```zig
const myFunc = std.someModule.someFunc;
```

Single letters are allowed:

```zig
fn example() void {
    const x = 10;
    _ = x;
}
```

## Rationale

Zig uses different cases to distinguish different kinds of identifiers:

- `snake_case` for variables, constants, and function parameters
- `PascalCase` for types
- `camelCase` for functions

This rule enforces snake_case for variables while allowing PascalCase for type aliases, which are detected by analyzing the assigned value.
