---
rule: Z011
title: Deprecated function call
enabled: true
---

# Z011: Deprecated function call

Detects calls to functions or methods marked as deprecated with `@"deprecated: ..."` doc comments.

## Examples

### Bad

```zig
const MyType = struct {
    value: u32 = 0,
    /// Deprecated: use newMethod instead
    pub fn oldMethod(self: *@This()) void {
        _ = self;
    }
};
const instance: MyType = .{};
// expect: Z011
pub fn main() void {
    instance.oldMethod();
}
```

### Good

Use the non-deprecated replacement:

```zig
const MyType = struct {
    value: u32 = 0,
    pub fn newMethod(self: *MyType) void {
        _ = self;
    }
};
var instance: MyType = .{};
pub fn main() void {
    instance.newMethod();
}
```

## Rationale

Deprecated functions are marked for removal in future versions. Using them creates technical debt and potential future breakage. This rule helps you find and update deprecated API usage before it becomes a problem.

The rule detects:
- Direct function calls to deprecated functions
- Method calls on types with deprecated methods
- Aliases to deprecated functions
- Deprecated type functions

Note: This rule requires semantic analysis and may not detect all deprecated calls when type information is unavailable.
